### DISCUSSION TITLE
GOAL: Derived Class Use Customizations for Base Class

### DISCUSSION THREAD

**Discussion by mwasson74** [AutoFixture/AutoFixture#1335] [Q&A] [answered]
I have scoured the interwebs and I can't seem to find an example that lets the base class customizations apply to each and every derived class. I see lots of examples on how to _omit_ specific properties but I don't want to do that.

Pseudo Code:

```csharp
fixture.Customize<BaseEntity>(c => c.With(e => e.IsDeleted, false));
var derived = fixture.Create<DerivedClass>();

Assert.False(derived.IsDeleted);
```

**Comment by aivascu** [comment, marked as answer]
@mwasson74 there is no "nice" way of doing it using the postprocess composer (aka `.Customize<>()`) however you can always create a builder tree that does what you require.

```cs
var node = new FilteringSpecimenBuilder(
    new FixedBuilder(false),
    new AndRequestSpecification(
        new PropertySpecification(typeof(bool), nameof(MyBaseClass.IsDeleted)),
        new MemberOwnerSpecification(
            new BaseTypeSpecification(typeof(MyBaseClass)))));
```
Note that I created two custom request specifications for this tree `MemberOwnerSpecification` that helps identify the owner of the currently processed member and `BaseTypeSpecification` which validates the type as a child type of a given type.
```cs
public class MemberOwnerSpecification : IRequestSpecification
{
    public MemberOwnerSpecification(IRequestSpecification typeSpecification)
    {
        TypeSpecification = typeSpecification
            ?? throw new ArgumentNullException(nameof(typeSpecification));
    }

    public IRequestSpecification TypeSpecification { get; }

    public bool IsSatisfiedBy(object request)
    {
        bool result = request switch
        {
            ParameterInfo parameter => TypeSpecification.IsSatisfiedBy(parameter.Member.ReflectedType),
            PropertyInfo property => TypeSpecification.IsSatisfiedBy(property.ReflectedType),
            FieldInfo field => TypeSpecification.IsSatisfiedBy(field.ReflectedType),
            _ => false
        };

        return result;
    }
}

public class BaseTypeSpecification : IRequestSpecification
{
    public BaseTypeSpecification(Type baseType)
    {
        BaseType = baseType;
    }

    public Type BaseType { get; }

    public bool IsSatisfiedBy(object request)
        => request is Type type
        && this.BaseType.IsAssignableFrom(type);
}
```
Now given that you have two types `MyBaseClass` that declares the `IsDeleted` property and `MyChildClass` that inherits `MyBaseClass`, then your test would look something like this.

```cs
[Fact]
public void AllGeneratedItemsAreDeleted()
{
    var node = new FilteringSpecimenBuilder(
        new FixedBuilder(false),
        new AndRequestSpecification(
            new PropertySpecification(typeof(bool), nameof(MyBaseClass.IsDeleted)),
            new MemberOwnerSpecification(
                new BaseTypeSpecification(typeof(MyBaseClass)))));

    var fixture = new Fixture();
    fixture.Customizations.Add(node);

    var items = fixture.CreateMany<MyChildClass>(10);

    var areAllFalse = items.All(x => x.IsDeleted == false);

    Assert.True(areAllFalse);
}
```
Now with a bit more code you could encapsulate that logic in a customization and reuse it in various scenarios.
Let me know if this helps.

**Reply by mwasson74** [reply]
@aivascu, this helps so much, thank you! 🙏

**Reply by MisinformedDNA** [reply]
Where are you setting `IsDeleted` to true? This `new PropertySpecification(typeof(bool), nameof(MyBaseClass.IsDeleted))` just identifies that the property exists, correct?

**Reply by MisinformedDNA** [reply]
Also, do you need to check for the existence of the property, if the base class is guaranteed to have it?

**Reply by aivascu** [reply]
> Where are you setting `IsDeleted` to true? This `new PropertySpecification(typeof(bool), nameof(MyBaseClass.IsDeleted))` just identifies that the property exists, correct?

In this particular case the property was set to `false`. Notice the `FixedBuilder`.

Also you don't need to check if the property exists. If a request will match the specification then it will resolve to the value in the specimen builder. If none will match then this builder is skipped entirely.

Guess you could say that the nameof expression is doing the validation in a way.
